src/cursor.cpp

```diff
@@ -661,88 +777,8 @@ static PyObject* execute(Cursor* cur, PyObject* pSql, PyObject* params, bool ski
-            return RaiseErrorFromHandle(cur->cnxn, "SQLParamData", cur->cnxn->hdbc, cur->hstmt);
-
-        TRACE("SQLParamData() --> %d\n", ret);
-
-        if (ret == SQL_NEED_DATA)
-        {
-            szLastFunction = "SQLPutData";
-            if (PyBytes_Check(pInfo->pObject))
-            {
-                const char* p = PyBytes_AS_STRING(pInfo->pObject);
-                SQLLEN offset = 0;
-                SQLLEN cb = (SQLLEN)PyBytes_GET_SIZE(pInfo->pObject);
-                while (offset < cb)
-                {
-                    SQLLEN remaining = min(pInfo->maxlength, cb - offset);
-                    TRACE("SQLPutData [%d] (%d) %.10s\n", offset, remaining, &p[offset]);
-                    Py_BEGIN_ALLOW_THREADS
-                    ret = SQLPutData(cur->hstmt, (SQLPOINTER)&p[offset], remaining);
-                    Py_END_ALLOW_THREADS
-                    if (!SQL_SUCCEEDED(ret))
-                        return RaiseErrorFromHandle(cur->cnxn, "SQLPutData", cur->cnxn->hdbc, cur->hstmt);
-                    offset += remaining;
-                }
-            }
-#if PY_VERSION_HEX >= 0x02060000
-            else if (PyByteArray_Check(pInfo->pObject))
-            {
-                const char* p = PyByteArray_AS_STRING(pInfo->pObject);
-                SQLLEN offset = 0;
-                SQLLEN cb     = (SQLLEN)PyByteArray_GET_SIZE(pInfo->pObject);
-                while (offset < cb)
-                {
-                    SQLLEN remaining = min(pInfo->maxlength, cb - offset);
-                    TRACE("SQLPutData [%d] (%d) %.10s\n", offset, remaining, &p[offset]);
-                    Py_BEGIN_ALLOW_THREADS
-                    ret = SQLPutData(cur->hstmt, (SQLPOINTER)&p[offset], remaining);
-                    Py_END_ALLOW_THREADS
-                    if (!SQL_SUCCEEDED(ret))
-                        return RaiseErrorFromHandle(cur->cnxn, "SQLPutData", cur->cnxn->hdbc, cur->hstmt);
-                    offset += remaining;
-                }
-            }
-#endif
-#if PY_MAJOR_VERSION < 3
-            else if (PyBuffer_Check(pInfo->pObject))
-            {
-                // Buffers can have multiple segments, so we might need multiple writes.  Looping through buffers isn't
-                // difficult, but we've wrapped it up in an iterator object to keep this loop simple.
-
-                BufferSegmentIterator it(pInfo->pObject);
-                byte* pb;
-                SQLLEN cb;
-                while (it.Next(pb, cb))
-                {
-                    Py_BEGIN_ALLOW_THREADS
-                    ret = SQLPutData(cur->hstmt, pb, cb);
-                    Py_END_ALLOW_THREADS
-                    if (!SQL_SUCCEEDED(ret))
-                        return RaiseErrorFromHandle(cur->cnxn, "SQLPutData", cur->cnxn->hdbc, cur->hstmt);
-                }
-            }
-#endif
-            ret = SQL_NEED_DATA;
-        }
-    }
+    if (!ProcessDAEParams(ret, cur, false))
+        return 0;
```

Comment:
Since the previous error handler on line 776 freed the parameter data, I bet we should do so too.  Unless it does it in PDP, but then the next line would be confusing.